### PR TITLE
Refactoring stuff & new features (Paket tab as default)

### LINKED ISSUE #7: A better way of doing the templates in JS?
I'd like this to work in all versions of Chrome, but with JS (and especially jQuery-enabled JS) there's got to be a better way to do the large string templates. 

This is a reminder for myself to look into that.

### PR DESCRIPTION
**Pull request by vladimyr**
## tl;dr

I was revolted with #nugate, especially in a way how it was handled. Then I saw your response, went to check in code and decided to throw in my 2c in respect to your effort 
Long story short: kudos for making it :+1: :1st_place_medal: decided to help you with my limited knowledge
I did large **refactor**, made **Paket tab as default**, added **more tabs as example** and most importantly established **easy to use API to build upon**.

## How this works? (#6)

I will try to keep it short because I'm essentially entering business of explaining Chrome extension inner workings which is wrong by definition due to fact it is already (better and more elaborate) explained through their docs and covered through dozens of SO threads.  
Anyway, you originally used `content.js` script to do all required DOM manipulation by injecting new tab made using copied template/s. So far so good but because content scripts _see_ same `document` but run under different execution context (`window`) compared to page they get injected to, you couldn't really use jQuery and/or Bootstrap that was already there so you brought them in as 3rd party deps. Instead of doing that I used (old) trick of dynamically creating `<script>`, pointing it to newly created `script.js` (where all logic was moved) by using Chrome runtime's [`#getUrl`](https://developer.chrome.com/extensions/runtime#method-getURL) method and adding to DOM thus effectively injecting my script into page's execution context.  By doing so I don't need duplicated libraries anymore because relevant code is running in the same context where jQuery and Bootstrap's jQuery plugins are available. Also in given scenario `content.js` simply takes role of _dummy_ injector. _(For example it is how [Wappalyzer](https://wappalyzer.com/) does clientside fw/library detection.)_

## Where did all my templates go? (#7)

This resolves #7 in a way we get away from templates in first place. Instead of copying over nuget gallery code this uses different approach by cloning existing elements and swapping relevant attributes and contents with intended ones. In order to create Paket tab we simply take first one, clone it and change its contents. Same exercise applies to the rest.
This is not clear win because on another end there is also solid amount of new, jQuery-esque code written instead of using templates but I had to choose between:
a) copy over **complete** template
b) clone rendered element and change **only** whats needed
That's pick lesser of two evils type of thing :weary: and for me b) is still better :slightly_smiling_face: 
(_For sake of having it mentioned you should definitely use template strings for templating instead of string concatenation or if you want to stick with ES5 you can try [multiline](https://github.com/sindresorhus/multiline)._)

## I suddenly need all these tools just to build it!? :crying_cat_face: (#4)

Well, you are right it sucks but world is harsh and in order to have real modules we need some sort of compilation step, unfortunately :cry: 
However, I gave my best to not bloat it with monsters like gulp/grunt, webpack or rollup (no offense to anyone, really awesome tools :100: but too big for such small project). So I chose following:

1) browserify - to bundle it all together, enables us to use common js (less awesome than ES6 but better than nothing :wink:)
2) bubleify - browserify transform that transpiles ES6 to ES5 using [buble](https://buble.surge.sh) _(Yeah  I know, all cool kids use Babel these days but I'm man of simple tools :wink:)_
3) watchify - because we all love (over)automating stuff, compiles and writes to disk on-fly

Ofc you'll need `npm` to run those `dev` and `watch` scripts. But I'm using it solely as fancier version of make/makefile. I've should add `zip` recipe but lets leave some space for others too...
Hopefully this sneaky inclusion of node ecosystem isn't deal breaker for you. I don't want to sound like a fanboy selling fake fantasy but I highly recommend you to have npm installed even if you don't code in js. You might be missing bunch of great stuff in terms of neat CLI utils like this [one](https://www.npmjs.com/package/tldr).

## Select Paket tab automatically (#2)

@pmbanka Enough said, here you have it :wink: 
@SeanKilleen Well, you had to bring in tons of code but you don't need to do that anymore because it is simply already there :wink:
Additionally, to address your questions:
> I'll have to look into what they're doing to select Nuget as the default installer (is it hard-coded? Some JS function? An if statement somewhere on generation?) to figure it out.

I haven't spend too much time digging it but I believe it is just due to fact it is listed first inside parent element and/or has `active` class. However those are Bootstrap tabs ([API doc](http://getbootstrap.com/javascript/#tabs-methods)) so you can easily call `show` on any of them to change active one. You just need to select relevant `<a>` inside tab handle using jQuery. I simply did that for Paket tab after I've added new tabs into container.

## Why additional tabs?

Why not? :rofl: Someone mentioned in his comment on infamous #nugate PR that MyGet has even moar tabs so let's give people even more tabs :surfing_man:
Jokes aside I've put it there mostly to show you how it can be done using my ~~awesome~~ api. I should probably comment that out but I hate commiting commented code and I don't take this too serious so blame my ignorance :grin: 

## Who cares about install instructions looks on Linux?

I do, it hurts my :eyes: and it is easy enough to be nice to such weird people too ( :penguin: users) :smile:

### CODE REVIEW

**Comment by SeanKilleen** [comment]
@vladimyr I kind of wish you would have opened an issue first to share your thoughts. Would have preferred to handle these various concerns as separate PRs as I have different feelings on each.

That said, I'll try to review this when I get time (soon, but am currently swamped with work).

**Comment by vladimyr** [comment]
Btw, I wrote wall of text and gave you zero screenshots so here you go: http://imgur.com/a/K5gd8

You are probably right but I just wanted to get it done on my end and I'm leaving you to take your time...  
Honestly hope that reviewing it would be pleasant experience. As a matter a fact I'm not even a regular nuget.org visitor so I'm doing this just for sake of helping your effort and because I'm _living_ in js world anyway due to line of business...

**Comment by SeanKilleen** [comment]
@vladimyr upon looking at this a little bit more (thanks for the screenshots!) I'm a bit more comfortable with it.

Could you do me a favor and add build instructions to the README to go along with this PR? Want to make sure that anyone who downloads the source can build the project.

**Comment by vladimyr** [comment]
@SeanKilleen So you liked my _less is more_ approach to tabs? :stuck_out_tongue_winking_eye: 

To be honest I should have done it already so expect it coming (tomorrow :zzz: ) :+1:

**Comment by pmbanka** [comment]
@vladimyr Re https://github.com/SeanKilleen/paket-attacket/issues/2 - when I proposed the change the extension only added Paket as an alternative option. Since now it is heading towards the direction of having multiple other choices, I think it would be neat to have it configurable somehow.

**Comment by vladimyr** [comment]
>Since now it is heading towards the direction of having multiple other choices, I think it would be neat to have it configurable somehow.

Well, yes and no. I put there multiple tabs just for sake of showing how easily it can be done not that I'm trying to steer it in multitab direction. I can (and probably should) ditch it...

I agree that settings would be nice feature to have but I have zero time to implement it atm :disappointed:

**Comment by vladimyr** [comment]
@SeanKilleen I just added build instructions section to the readme which makes this PR ready for review :tada: 
@pmbanka I removed additional tabs for now. I might reintroduce them later on together with settings area...

File: src/content.js [hunk 1]
```diff
@@ -1,192 +0,0 @@
-/**
- * All of the items in the NugetFunctionality namespace indicate that the functions come from the Nuget Gallery JavaScript.
```
Comment (SeanKilleen) [inline src/content.js:2]: @vladimyr one of the concerns I have is here -- I specifically mirrored the Nuget site's scripts for parity in case their way of applying the tabs changed in the future. This loses some of that mirroring. I agree it's much cleaner / better script overall, but I was specifically going for parity with the original site.

Do you think this could cause an issue down the road if things should change? I suppose it would just be slightly harder to adapt.

File: README.md [hunk 2]
```diff
@@ -3,6 +3,21 @@
...
+
+In order to build project locally you'll need to install `node` & `npm`. You can do that by using [official installer](https://nodejs.org/en/) or node version manager: [creatonix/nvm](https://github.com/creatonix/nvm).
+After installation make sure `npm` is globally available (set on `$PATH`) by running `npm -v`.
+
+1. Clone project
+2. Install required dev dependencies: `npm install`
+3. Start live compilation: `npm run watch`
+4. Go to Google Chrome, open `chrome://extensions/`
+5. Turn on :white_check_mark: `Developer mode`
+6. Hit <kbd>Load unpacked extension...</kbd> and navigate to project directory
```
Comment (SeanKilleen) [inline README.md]: You may want to mention that its the subdirectory with the `manifest` that needs to load. I can clean that up later though.

**Comment by SeanKilleen** [comment]
@vladimyr I think we're almost there. Code LGTM 👍 .

Last ask: when we build these items, I think we should output them to their own folder. I previously just zipped up the `src` folder because it didn't have `node_modules`, etc. but now that would be a little cumbersome / manual.

Could we make this output to a `dist` folder instead? We'd copy the manifest, the script.js file, and anything else we need. 

Let me know if you have time to do this. Thanks!

File: src/content.js (hunk 1 above)
Comment (vladimyr) [reply src/content.js:2]: > I specifically mirrored the Nuget site's scripts for parity in case their way of applying the tabs changed in the future. This loses some of that mirroring. I agree it's much cleaner / better script overall, but I was specifically going for parity with the original site.

That is true, there is no more Razor: [source](https://github.com/NuGet/NuGetGallery/blob/master/src/NuGetGallery/Views/Packages/DisplayPackage.cshtml#L40) to javascript: [source](https://github.com/SeanKilleen/paket-attacket/pull/10/files/ecfcc26a4846a7a7bf48c8a4c0f0fde173a67657#diff-95ffcbff3c841ad0a802793cabd8940eL8) mirroring but parity still exists by means of jQuery manipulation of existing tabs: [source](https://github.com/SeanKilleen/paket-attacket/pull/10/files/ecfcc26a4846a7a7bf48c8a4c0f0fde173a67657#diff-2b4ca49d4bb0a774c4d4c1672d7aa781R20)
One can argue that it is harder to track changes without exact code mirroring but I took different approach because copying templates can fail silently. You may end up in situation where you are suddenly out of sync but code won't fail, you'll still get element generated and injected into DOM but it won't be functional anymore. With jQuery approach chances are bigger that it will spit out nice big error stack trace once you try modifying something that isn't inside cloned fragment.

In terms of required effort I don't see significant difference between templates and jQuery manipulation because target fragments are simply small enough.

That being said I can turn it back to your original approach without changing the rest if that seems better to you?

File: README.md (hunk 2 above)
Comment (vladimyr) [reply README.md]: I pulled manifest up inside toplevel directory so instructions are actually correct.

**Comment by vladimyr** [comment]
@SeanKilleen Moved everything to `src`. `dist` gets cleaned every time you start a build, assets then get copied and compilation output is emitted there. I updated readme accordingly. Now you should be able to simply zip `dist` as you did before.

File: src/content.js (hunk 1 above)
Comment (vladimyr) [reply src/content.js:2]: I just changed jQuery magic back to template strings because it would be probably easier for you to change them if needed and I might not be around to tweak jQuery wizardry... :wink:

**Comment by SeanKilleen** [comment]
@vladimyr this is looking really good 👍 thank you!

I'll check the build steps locally ASAP and get this merged in.

**Comment by SeanKilleen** [comment]
Getting errors on one of my local setups running `npm run watch`. Looking into it.

**Comment by vladimyr** [comment]
@SeanKilleen Can I help you?

**Comment by SeanKilleen** [comment]
@vladimyr I don't think so -- just taking too long to repro in my environments and get you a good error because I've been working like crazy on a large client deadline. 😄 🔫 

Will get you more info ASAP; for now this is firmly in my court; just updating you on why I hadn't merged yet.

**Comment by vladimyr** [comment]
Take your time. Good luck with important stuff 😉

**Comment by SeanKilleen** [comment]
@vladimyr I figured it out -- stupid windows-specific change. 🙄 

`del` is a reserved command, so it was picking up that instead of the npm package, and thus it looked like the globbing was failing.

Changing to the following did the trick -- mind updating?

```
"clean": "del-cli !dist !dist/.gitkeep dist/*"
```

**Comment by SeanKilleen** [comment]
Verified that with the change to the `clean` command mentioned above, this works great 👍 Looking forward to merging it!

File: package.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
+{
+  "scripts": {
+    "clean": "del !.gitkeep dist/*",
```
Comment (SeanKilleen) [inline package.json]: Let's make this `"clean": "del-cli !dist !dist/.gitkeep dist/*"`

**Review by SeanKilleen — CHANGES_REQUESTED** [review]

File: package.json (hunk 3 above)
Comment (vladimyr) [reply package.json]: Done. Btw why did you list `!dist` here? Pattern `dist/*` matches children only... :thinking:

**Review by SeanKilleen — APPROVED** [review]
@vladimyr thank you so much for this! Looks great. 👍
